flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java

```diff
@@ -243,7 +249,23 @@ void requestPartitions() throws IOException, InterruptedException {
 				}
 
 				for (InputChannel inputChannel : inputChannels.values()) {
-					inputChannel.requestSubpartition();
+					executor.submit(() -> {
+						try {
+							inputChannel.initializeState(reader);
+						} catch (Throwable t) {
+							inputChannel.setError(t);
+						}
+					});
+				}
+
+				for (InputChannel inputChannel : inputChannels.values()) {
+					executor.submit(() -> {
+						try {
+							inputChannel.requestSubpartition();
```

Comment:
Could it be that we start reading from a channel before we actually request subpartition? For example:
```
Optional<BufferAndAvailability> RemoteInputChannel#getNextBuffer() throws IOException {
	(...)
	checkState(partitionRequestClient != null, "Queried for a buffer before requesting a queue."); // ??
```

or what if `partitionRequestClient` is already not null, but 
```
partitionRequestClient.requestSubpartition(partitionId, subpartitionIndex, this, 0);
```
hasn't completed?

Reply:
Yes, it has another issue here.  In order to compatible with previous process, we can divide the previous `requestSubpartition` into two steps. The first step is to create client which can be done during `InputGate#setup` as before. The second step is to actual request partition which is triggered after reading channel state as done in this PR change. 

To do so, this check is still valid, but I guess it is still breaking the previous assumption which actually wants to guarantee the partition is really requested before calling `getNextBuffer`. But now the existing client does not mean the partition was requested already. So another option might remove this `checkState`.

Reply:
I currently remove this `checkState` in PR for simple. If you have other thoughts, then we can further sync afterwards.